keras_cv/losses/giou_loss.py

```diff
@@ -0,0 +1,151 @@
+# limitations under the License.
+
+
+import tensorflow as tf
+
+from keras_cv import bounding_box
+
+
+class GIoULoss(tf.keras.losses.Loss):
+    """Implements the GIoU Loss
+    GIoU loss is a modified IoU loss commonly used for object detection. This loss aims
+    to directly optimize the IoU score between true boxes and predicted boxes. GIoU loss
+    adds a penalty term to the IoU loss that takes in account the area of the
+    smallest box enclosing both the boxes being considered for the iou. The length of
+    the last dimension should be atleast 4 to represent the bounding boxes. While
+    this dimension can have more than 4 values, these values will be ignored for the
+    calculation of this loss.
+    Args:
+        bounding_box_format: a case-insensitive string (for example, "xyxy").
+            Each bounding box is defined by at least these 4 values. The inputs
+            may contain additional information such as classes and confidence after
+            these 4 values but these values will be ignored while calculating
+            this loss. For detailed information on the supported formats, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+        axis: the axis along which to mean the ious. Defaults to `None` which implies
+            mean across no axes.
+
+    References:
+        - [GIoU paper](https://arxiv.org/pdf/1902.09630)
+        - [TFAddons Implementation](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/GIoULoss)
+    Sample Usage:
+    ```python
+    y_true = tf.random.uniform((5, 10, 5), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    y_pred = tf.random.uniform((5, 10, 4), minval=0, maxval=10, dtype=tf.dtypes.int32)
+    loss = GIoULoss(bounding_box_format = "xyWH")
+    loss(y_true, y_pred).numpy()
+    ```
+    Usage with the `compile()` API:
+    ```python
+    model.compile(optimizer='adam', loss=keras_cv.losses.GIoULoss())
+    ```
+    """
+
+    def __init__(self, bounding_box_format, axis=None, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.axis = axis
+
+    def _compute_giou(self, boxes1, boxes2):
+        boxes1_rank = len(boxes1.shape)
+        boxes2_rank = len(boxes2.shape)
+
+        if (
+            boxes1_rank != boxes2_rank
+            or boxes1_rank not in [2, 3]
+            or boxes2_rank not in [2, 3]
+        ):
+            raise ValueError(
+                "`GIoULoss` expects both boxes to be batched, or both "
+                f"boxes to be unbatched.  Received `len(boxes1.shape)`={boxes1_rank}, "
+                f"`len(boxes2.shape)`={boxes2_rank}.  Expected either `len(boxes1.shape)`=2 AND "
+                "`len(boxes2.shape)`=2, or `len(boxes1.shape)`=3 AND `len(boxes2.shape)`=3."
+            )
+
+        if self.bounding_box_format.startswith("rel"):
+            target = "rel_yxyx"
```

Comment:
any reason we cannot always use a single target format?

Reply:
This is mostly to remove dependency on images for this loss.

For example if the passed bbox format is non-rel and we use only rel for this function, the images will be required for the dimension calculation (discussed in #604 and fixed for `compute_iou` in #612)

Reply:
I think that this is automatic in convert_format() now, right?

Reply:
I don't think it is. `convert_format` just has support for rel->rel conversions without an image.
Earlier, even rel->rel conversion required images because the intermediate conversion was xyxy.

This part ensures that rel source has rel targets and the (same for non-rel ones).

Reply:
Ah I see; ok that makes sense.  I bet we can write a utility for this.

Something like:
```
preserve_rel('xywh', bounding_box_format=bounding_box_format)
```

Returns 'rel_xywh' if bounding_box_format starts with rel, otherwise 'xywh'.

Reply:
Although I think the current way is better, I have opened a PR for this (#1016).

Lmk if this is what you were looking for and I will rebase this one and add it to `_compute_giou` as well! @LukeWood